src/rtcModule/webrtcAdapter.cpp

```diff
@@ -78,171 +88,382 @@ unsigned long generateId()
-        webrtc::VideoCaptureFactory::CreateDeviceInfo());
-    if (!info)
+absl::optional<bool> CaptureModuleLinux::needs_denoising() const
+{
+    return absl::nullopt;
+}
+
+bool CaptureModuleLinux::GetStats(webrtc::VideoTrackSourceInterface::Stats *stats)
+{
+    return false;
+}
+
+void CaptureModuleLinux::AddOrUpdateSink(rtc::VideoSinkInterface<webrtc::VideoFrame>* sink, const rtc::VideoSinkWants& wants)
+{
+    RTC_DCHECK(mWorkerThreadChecker.IsCurrent());
+    mBroadcaster.AddOrUpdateSink(sink, wants);
+}
+
+void CaptureModuleLinux::RemoveSink(rtc::VideoSinkInterface<webrtc::VideoFrame>* sink)
+{
+    RTC_DCHECK(mWorkerThreadChecker.IsCurrent());
+    mBroadcaster.RemoveSink(sink);
+}
+
+bool CaptureModuleLinux::remote() const
+{
+    return mRemote;
+}
+
+void CaptureModuleLinux::OnFrame(const webrtc::VideoFrame& frame)
+{
+    mBroadcaster.OnFrame(frame);
+}
+
+void CaptureModuleLinux::AddRef() const
+{
+    mRefCount.IncRef();
+}
+
+rtc::RefCountReleaseStatus CaptureModuleLinux::Release() const
+{
+    const auto status = mRefCount.DecRef();
+    if (status == rtc::RefCountReleaseStatus::kDroppedLastRef)
     {
-        RTCM_LOG_WARNING("Could not enumerate video devices. Using the default device.");
-        mInputDevices.video.push_back(cricket::Device("default", "0"));
-        return;
+        delete this;
     }
 
-    int numDevices = info->NumberOfDevices();
-    RTCM_LOG_INFO("Enumerate input devices: %d found.", numDevices);
-    auto& devices = mInputDevices.video;
-    for (int i = 0; i < numDevices; ++i)
+    return status;
+}
+
+std::set<std::pair<std::string, std::string> > CaptureModuleLinux::getVideoDevices()
+{
+    std::set<std::pair<std::string, std::string>> videoDevices;
+    std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> info(webrtc::VideoCaptureFactory::CreateDeviceInfo());
+    if (!info)
     {
-        const uint32_t kSize = 256;
-        char name[kSize] = {0};
-        char id[kSize] = {0};
-        if (info->GetDeviceName(i, name, kSize, id, kSize) != -1)
-        {
-            std::string sId = id;
-            if (sId.find(":1") == std::string::npos)
-            {
-                devices.push_back(cricket::Device(name, id));
-            }
-            else
-            {
-                devices.insert(devices.begin(), cricket::Device(name, id));
-            }
-            RTCM_LOG_INFO("Video input device %d: '%s'", i+1, name);
-        }
+        return videoDevices;
     }
 
-    if (devices.size() == 0)
+    uint32_t numDevices = info->NumberOfDevices();
+    for (uint32_t i = 0; i < numDevices; i++)
     {
-        RTCM_LOG_WARNING("Not suitable input video devices were found. Default device added.");
-        devices.push_back(cricket::Device("default", "0"));
+        char deviceName[256];
+        char uniqueName[256];
+        std::unique_ptr<webrtc::VideoCaptureModule::DeviceInfo> deviceInfo(webrtc::VideoCaptureFactory::CreateDeviceInfo());
```

Comment:
Cannot we use the `info` retrieved previously and simply iterate through the items and call `info->GetDeviceName()`??